### ISSUE TITLE
Question about how GPT learns to be "generative"

### ISSUE THREAD

**Issue by metalwhale** [karpathy/nanoGPT#432]
I'm investigating deeply the reason why GPTs are referred to as "generative" models. I've found this repository to be a very helpful resource. Thank you so much for publishing it!

As the title suggests, I'm confused about the way GPTs are trained. AFAIK to be called "generative," a model should have the ability to learn the joint probability over the training data. In the case of LLMs, we apply the chain rule of Bayes' formula to achieve this by leveraging the autoregressive method for every token of each input text sequence. For example, with a text sequence of 4 tokens, it can be written as:
```
p(x4,x3,x2,x1) = p(x4|x3,x2,x1) * p(x3|x2,x1) * p(x2|x1) * p(x1)
```
where `x1` denotes the 1st token, `x2` denotes the 2nd token and so on, respectively.

I understand the conditional terms `p(x_n|...)` where we use cross-entropy to calculate their losses. However, I'm unsure about the probability of the very first token `p(x1)`.

How is it calculated? Where does it appear in the source code? Is it in some configurations of the training process, or in the model architecture, or in the loss function? Can you enlighten me?

IMHO, if the model doesn't learn `p(x1)` properly, the entire formula for Bayes' rule cannot be completed, and we can't refer to GPTs as "truly generative". Am I missing something here?

Thank you in advance.

**Comment by relic-yuexi** [comment]
GPTs (Generative Pre-trained Transformers) are indeed generative models, and they are trained to predict the next token in a sequence given the previous tokens. The generative aspect comes from the model's ability to learn the conditional probability distribution of the next token (or tokens) given the context of the previous tokens.

In the context of GPTs, the model is trained using a technique called "masked language modeling" (MLM), which is a form of autoregressive language modeling. During training, a portion of the input tokens are masked (replaced with a special token, often represented as "[MASK]"). The model's task is to predict the masked tokens based on the context provided by the surrounding unmasked tokens.

The probability of the first token, p(x1), is not explicitly learned in the same way as the conditional probabilities p(x_n|x1, ..., x(n-1)). Instead, it is implicitly learned as part of the model's initial state. The model's parameters, including the embeddings and weights, are initialized randomly and then fine-tuned during training. The initial token (x1) is treated like any other token in the sequence, and the model learns to predict it based on the context of the preceding tokens, which in the case of the first token, is typically an artificial context like a start-of-sequence (SOS) token or a series of zeros.

In practice, the model learns to generate coherent text by predicting each token one at a time, starting from the beginning of the sequence. The model's architecture, particularly the Transformer's attention mechanism, allows it to consider the entire context when making predictions, which enables it to generate text that is coherent and contextually relevant.

The loss function used during training is typically cross-entropy, which measures the difference between the predicted probability distribution over the next token and the actual distribution (the one-hot encoding of the true next token). This loss is calculated for each token in the sequence, and the model is trained to minimize this loss.

In summary, while GPTs do not explicitly learn the probability of the first token, they learn to generate text that is coherent from the start by predicting each subsequent token based on the context. The model's ability to generate text from scratch is what makes it generative, even though the initial token's probability is not directly calculated. The model's architecture and training process ensure that it can generate text that follows the patterns and structures of the language it was trained on.

**Comment by metalwhale** [comment]
@relic-yuexi 
Thank you for your very detailed answer! Can I ask you some questions?

> The model's task is to predict the masked tokens based on the context provided by the surrounding unmasked tokens.

AFAIK, GPTs use Causal Language Modeling (CLM) for training, which is not MLM. CLM is unidirectional, meaning that GPTs learn to predict the next tokens based on the previous tokens, rather than predicting masked tokens based on the surrounding unmasked tokens. Am I correct?

> The initial token (x1) is treated like any other token in the sequence, and the model learns to predict it based on the context of the preceding tokens, which in the case of the first token, is typically an artificial context like a start-of-sequence (SOS) token or a series of zeros.

How can I feed this "initial context" (as you mentioned, an SOS/BOS token or a series of zeros) to the model? Because, as I understand it, we don't have any way to ask the model to generate arbitrary texts without providing at least one token as the input prompt.

Thank you in advance.

**Comment by relic-yuexi** [comment]
You are right. CausalSelfAttention use `att = att.masked_fill(self.bias[:,:,:T,:T] == 0, float('-inf'))`.

"initial context" will be in the make_datasets.

If i have a sample sentence like "i like the cat ", it will be maked as "<bos>i like the cat<eos>". Then it will be encode as some numers.

Actually you can use none token to generate, it is a greedy method.

You can learn some knowledge at "[Let's build GPT: from scratch, in code, spelled out.](https://youtu.be/kCc8FmEb1nY?si=ibvxqaOF1t5Tn1QR)".

**Comment by metalwhale** [comment]
@relic-yuexi Thank you for your advice! Will definitely take a look!

About this repository in particular, forgive my selfish mention but @karpathy could you please answer my question when you have time. Thank you so much!

**Comment by metalwhale** [comment]
I asked the same question at https://github.com/huggingface/transformers/issues/28860, feel free to join
